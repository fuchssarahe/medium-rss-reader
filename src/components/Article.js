import React, { Component } from 'react';

class Article extends Component {
  render() {
    const item = this.props.item;
    return (
      <li>
        <a href={item.link}>
        <h3>{item.title}</h3>
        <h4>{item.description} | {item.creator} | {item.pubDate}</h4>
        <h4>Categories: {item.categories.join(' | ')}</h4>
        </a>
      </li>
    );
  }
}

export default Article;
