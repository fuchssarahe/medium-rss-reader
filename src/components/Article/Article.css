.Article {
  border: 1px solid black;
  margin: 1em auto;
  max-width: 800px;
  padding: 0.5em;
  width: 80%;
}

.Article__title {
  font-weight: bold;
  padding-bottom: 0.5em;
}

.Article__metadata {
  margin-left: 0.5em;
}

.Article__snippet {
  margin-left: 0.5em;
}
