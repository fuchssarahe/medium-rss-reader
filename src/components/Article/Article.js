import React from 'react';
import PropTypes from 'prop-types';
import Categories from '../Categories/Categories';
import './Article.css';

// Responsible for rendering the main content of an article.
const Article = ({ item, addTagFilter }) => {
  return (
    <li className='Article'>
      <Title url={item.link} title={item.title} />
      <Metadata creator={item.creator} date={item.pubDate} />
      {item.categories && <Categories categories={item.categories} onClick={addTagFilter} />}
      {item.contentSnippet && <Snippet text={item.contentSnippet} url={item.link} />}
    </li>
  );
}

Article.propTypes = {
  item: PropTypes.object.isRequired,
  addTagFilter: PropTypes.func.isRequired
}

const Title = ({ url, title }) => {
  return (
    <a href={url}>
      <h3 className='Article__title'>{title}</h3>
    </a>
  );
};

Title.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired
}

const Metadata = ({ creator, date }) => {
  return (
    <h4 className='Article__metadata'>{creator} | {date}</h4>
  );
}

Metadata.propTypes = {
  creator: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired
}

const Snippet = ({ url, text }) => {
  return (
    <p className='Article__snippet'>
      {formatSnippet(text)}
      <a href={url}>Read More >></a>
    </p>
  );
}

Snippet.propTypes = {
  url: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired
}

function formatSnippet(snippet) {
  const trimmedSnippet = snippet.slice(0,-34).trim();
  let suffix;
  if (trimmedSnippet[trimmedSnippet.length - 1] === '.') {
    suffix = ' ';
  } else {
    suffix = '... '
  }
  return trimmedSnippet + suffix;
}

export default Article;
