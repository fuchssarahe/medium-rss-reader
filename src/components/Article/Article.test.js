import React from 'react';
import { shallow } from 'enzyme';
import Article from './Article';

let userArticle = {
  creator: 'Alan Levine',
  title: 'Attribute always. Period.',
  link: 'https://medium.com/@cogdog/uuid',
  pubDate: 'Fri, 24 Aug 2018 17:41:22 GMT',
  'content:encoded': '<p>Attribute always. Period.</p><p>Even if it is p…ll_rss&postId=95b4d2f7880a'
}

let publicationArticle = {
  categories: ['cities', 'economics', 'retail', 'business', 'housing'],
  content: '"<div class="medium-feed-item"><p class="medium-feed-image"><a href="https://medium.com/the-atlantic/uuid"><img src="https://cdn-images-1.medium.com/max/2000/1*zWOz6wBx-x0zKt8ALrWRiw.jpeg" width="5861"></a></p><p class="medium-feed-snippet">New York&#x2019;s empty storefronts are a dark omen for the future of cities</p><p class="medium-feed-link"><a href="https://medium.com/the-atlantic/uuid">Continue reading on The Atlantic »</a></p></div>"',
  contentSnippet: 'New York’s empty storefronts are a dark omen for the future of citiesContinue reading on The Atlantic »',
  creator: 'Derek Thompson',
  'dc:creator': 'Derek Thompson',
  guid: 'https://medium.com/p/a59edb0399b8',
  isoDate: '2018-10-15T15:30:12.000Z',
  link: 'https://medium.com/the-atlantic/uuid',
  pubDate: 'Mon, 15 Oct 2018 15:30:12 GMT',
  title: 'What’s the Matter With Manhattan?',
}

describe('<Article />', () => {

  test('it handles a sparse user feed', () => {
    expect(shallow(<Article item={userArticle} addTagFilter={() => {}}/>)).toMatchSnapshot();
  });

  test('it handles a publication feed', () => {
    expect(shallow(<Article item={publicationArticle} addTagFilter={() => {}}/>)).toMatchSnapshot();
  });

  test('it does not render unknown html', () => {
    const component = shallow(<Article item={userArticle} addTagFilter={() => {}}/>);
    expect(component.html().indexOf('Even if it is ')).toBe(-1);
  })
});
