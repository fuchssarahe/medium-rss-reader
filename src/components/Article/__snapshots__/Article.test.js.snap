// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> it handles a publication feed 1`] = `
<li
  className="Article"
>
  <Title
    title="What’s the Matter With Manhattan?"
    url="https://medium.com/the-atlantic/uuid"
  />
  <Metadata
    creator="Derek Thompson"
    date="Mon, 15 Oct 2018 15:30:12 GMT"
  />
  <Categories
    categories={
      Array [
        "cities",
        "economics",
        "retail",
        "business",
        "housing",
      ]
    }
    onClick={[Function]}
  />
  <Snippet
    text="New York’s empty storefronts are a dark omen for the future of citiesContinue reading on The Atlantic »"
    url="https://medium.com/the-atlantic/uuid"
  />
</li>
`;

exports[`<Article /> it handles a sparse user feed 1`] = `
<li
  className="Article"
>
  <Title
    title="Attribute always. Period."
    url="https://medium.com/@cogdog/uuid"
  />
  <Metadata
    creator="Alan Levine"
    date="Fri, 24 Aug 2018 17:41:22 GMT"
  />
</li>
`;
